added merge functionality to javascript client sdk

PJGitLan opened the pull request.

thjaeckle commented:
Hello @PJGitLan and thanks for this contribution.
Will take a look once the "Checks" for this PR are all green.

For the "eclipsefdn/eca" check you will have to create an Eclipse account and sign the ECA (Eclipse Contributor Agreement), as also noted in the [CONTRIBUTING.md](https://github.com/eclipse-ditto/ditto-clients/blob/master/CONTRIBUTING.md#legal-requirements) - and sign your commit with the email address registered in your Eclipse account.

thjaeckle commented on javascript/lib/api/src/client/request-factory/request-sender.ts:
```diff
@@ -51,6 +51,19 @@ export abstract class RequestSender {
 
+  public fetchMergeRequest<T>(options: ParseRequest<T>): Promise<MergeResponse<T>> {
+    return this.fetchRequest(options).then(response => {
+      if (response.status === 204) {
```
As far as I remember a PATCH / merge could also create a new feature or property ..
So IMO checking for the status code 201 is missing here.

PJGitLan replied:
this should be resolved by using fetchPutRequest()

thjaeckle commented on javascript/lib/api/src/client/request-factory/request-sender.ts:
```diff
@@ -51,6 +51,19 @@ export abstract class RequestSender {
+    return this.fetchRequest(options).then(response => {
+      if (response.status === 204) {
+        return new MergeResponse(
+          options.parser(response.body),
```
For status 204 (no content) there is no content in the response, so parsing it is not required, you can pass in `null` (same as for PUT).

PJGitLan replied:
Went back to using PutResponse

thjaeckle commented on javascript/lib/api/src/client/request-factory/websocket-request-sender.ts:
```diff
@@ -95,7 +98,8 @@ export class WebSocketRequestSender extends RequestSender {
     const path = WebSocketRequestSender.buildPath(options.path);
-    const headers = this.buildHeaders(options.requestOptions, { 'content-type': 'application/json' });
+    const requestOptions = options?.requestOptions;
+    const headers = this.buildHeaders(requestOptions, requestOptions?.getHeaders().has('Content-Type') === true ? undefined : { 'Content-Type': 'application/json' });
```
👍

thjaeckle commented on javascript/lib/api/src/model/response.ts:
```diff
@@ -65,6 +65,27 @@ export class PutResponse<T> implements GenericResponse {
+    private readonly _headers: Map<string, string>
+  ) {}
+
+
```
`wasCreated` and `wasUpdated` (like for the PutResponse) could also be added.

PJGitLan replied:
Looks like putResponse en mergeResponse are the same so I removed mergeResponse and used putResponse instead.

thjaeckle reviewed (changes requested):
Thanks @PJGitLan for the contribution. 

I added some remarks / wishes for changing inline.
Good that you added "some" merge functionality to the JS SDK. 👍 

However, it would be nice if we go into that direction in the JS SDK to support PATCH / merge, that all levels are supported.
Do you think you can also address all other levels (where currently only "PUT" is available in the client)?
From the top of my head:
* merge Thing
* merge Thing attributes
* merge Thing attribute
* merge Features (all)
* merge Feature (single)
* merge Feature desired properties

That way the complete "merge" would be available in the JS client and not only a small selected subpart.

PJGitLan commented:
> Thanks @PJGitLan for the contribution.
> 
> I added some remarks / wishes for changing inline. Good that you added "some" merge functionality to the JS SDK. +1
> 
> However, it would be nice if we go into that direction in the JS SDK to support PATCH / merge, that all levels are supported. Do you think you can also address all other levels (where currently only "PUT" is available in the client)? From the top of my head:
> 
> * merge Thing
> * merge Thing attributes
> * merge Thing attribute
> * merge Features (all)
> * merge Feature (single)
> * merge Feature desired properties
> 
> That way the complete "merge" would be available in the JS client and not only a small selected subpart.

I added further merge functionality and solved previous remarks

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -142,6 +172,17 @@ export interface HttpThingsHandle extends ThingsHandle {
    */
   putPolicyId(thingId: string, policyId: string, options?: MatchOptions): Promise<PutResponse<string>>;
```
Any reason why policyId and definition are only used for http-requests and not for websocket-request?

thjaeckle commented on javascript/lib/api/src/client/constants/http-verb.ts:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum HttpVerb {
```
Nice that you added this enum 👍 
I however would have expected that the existing code is then also adjusted to use the enum. 
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

PJGitLan replied:
No problem, refactored the existing code

thjaeckle commented on javascript/lib/api/src/client/constants/header.ts:
```diff
@@ -0,0 +1,16 @@
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum Header {
```
Nice that you added this enum 👍
I however would have expected that the existing code is then also adjusted to use the enum.
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

PJGitLan replied:
done

thjaeckle commented on javascript/lib/api/src/client/constants/content-type.ts:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum ContentType {
```
Nice that you added this enum 👍
I however would have expected that the existing code is then also adjusted to use the enum.
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

PJGitLan replied:
done

thjaeckle reviewed (changes requested):
Thanks, that looks very promising 👍 
I added some remarks, mainly about the js-doc, which were copy&pasted without adjustment at some places.

I like the added enums, but we then already should use them everywhere to get the JS client in a better shape slowly :)

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
+   * merges an Attribute of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param attributePath - The new path to the new Attribute.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
+   *
+   * @param thingId - The ID of the Thing.
+   * @param attributePath - The new path to the new Attribute.
+   * @param attributeValue - The new Attribute.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/features.interfaces.ts:
```diff
@@ -89,6 +107,16 @@ export interface FeaturesHandle {
+   * Merges the definition of the specified Feature.
+   *
+   * @param featureId - The ID of the Feature.
+   * @param definition - The new definition.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/features.interfaces.ts:
```diff
@@ -110,6 +138,27 @@ export interface FeaturesHandle {
+   *
+   * @param featureId - The ID of the Feature.
+   * @param propertyPath - The path to the Property.
+   * @param property - The Property.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -123,6 +144,15 @@ export interface HttpThingsHandle extends ThingsHandle {
+  /**
+   * Adds or merges a Thing.
+   *
+   * @param thing - The Thing to create.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -142,6 +172,17 @@ export interface HttpThingsHandle extends ThingsHandle {
+   * Adds or updates the policyId of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param policyId - The new PolicyId.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
+   * Merges the Attributes of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param attributes - The new Attributes.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/features.interfaces.ts:
```diff
@@ -110,6 +138,27 @@ export interface FeaturesHandle {
+   * Merges the properties of the specified Feature.
+   *
+   * @param featureId - The ID of the Feature.
+   * @param properties - The properties.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/things.interfaces.ts:
```diff
@@ -161,6 +202,16 @@ export interface HttpThingsHandle extends ThingsHandle {
+   * merges the definition of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param definition - The new definition.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/features.interfaces.ts:
```diff
@@ -70,6 +70,24 @@ export interface FeaturesHandle {
+  /**
+   * Merges a Feature of this handle's Thing.
+   *
+   * @param feature - The new Feature.
```
done

PJGitLan commented on javascript/lib/api/src/client/handles/features.interfaces.ts:
```diff
@@ -70,6 +70,24 @@ export interface FeaturesHandle {
+  /**
+   * Merges all Features of this handle's Thing.
+   *
+   * @param features - The new Features.
```
done

thjaeckle reviewed (approved):
@PJGitLan the changes look good to me, very nice :+1:
Great that you did the extra changes to completely support "merge". Thanks you for the contribution.

PJGitLan commented:
> @PJGitLan the changes look good to me, very nice +1 Great that you did the extra changes to completely support "merge". Thank you for the contribution.

Hi @thjaeckle, no problem. Any idea when version 3.2.0 will be published?

thjaeckle commented:
Today 😉
